server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,242 @@
+            long diff = newRamUsage > defaultRamUsage ? newRamUsage - defaultRamUsage : defaultRamUsage - newRamUsage;
+            ramAccounting.addBytes(diff);
+            state = new TopKState(new ItemsSketch<>(maxMapSize), limit);
+        }
+        Object value = args[0].value();
+        state.itemsSketch.update(value);
+        return state;
+    }
+
+    private int maxMapSize(int x) {
+        // max map size should be 4 * the limit based on the power of 2 to avoid errors
+        return (int) Math.pow(2, Math.ceil(Math.log(x) / Math.log(2))) * 4;
+    }
+
+    private long calculateRamUsage(long maxMapSize) {
+        // The internal memory space usage of item sketch will never exceed 18 * maxMapSize bytes, plus a small
+        // constant number of additional bytes.
+        // https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+        return maxMapSize * 18L;
+    }
+
+    @Override
+    public TopKState reduce(RamAccounting ramAccounting, TopKState state1, TopKState state2) {
+        return new TopKState(state1.itemsSketch.merge(state2.itemsSketch), Math.min(state1.limit, state2.limit));
+    }
+
+    @Override
+    public List<Object> terminatePartial(RamAccounting ramAccounting, TopKState state) {
+        ItemsSketch.Row<Object>[] frequentItems = state.itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+        int limit = Math.min(frequentItems.length, state.limit);
+        var result = new ArrayList<>(limit);
+        for (int i = 0; i < limit; i++) {
+            var item = frequentItems[i];
+            var row = new HashMap<>(2);
+            row.put("item", item.getItem());
+            row.put("frequency", item.getEstimate());
+            result.add(row);
+        }
+        return result;
+    }
+
+    public DataType<?> partialType() {
+        return TopKStateType.INSTANCE;
+    }
+
+    public record TopKState(ItemsSketch<Object> itemsSketch, int limit) {
+    }
+
+    static final class TopKStateType extends DataType<TopKState> implements Streamer<TopKState> {
+
+        public static final int ID = 4232;
+        public static final TopKStateType INSTANCE = new TopKStateType();
+
+        @Override
+        public int id() {
+            return ID;
+        }
+
+        @Override
+        public Precedence precedence() {
+            return Precedence.CUSTOM;
+        }
+
+        @Override
+        public String getName() {
+            return "top_k_state";
+        }
+
+        @Override
+        public Streamer<TopKState> streamer() {
+            return this;
+        }
+
+        @Override
+        public TopKState sanitizeValue(Object value) {
+            return (TopKState) value;
+        }
+
+        @Override
+        public TopKState readValueFrom(StreamInput in) throws IOException {
+            int limit = in.readInt();
+            SketchStreamer<Object> streamer = new SketchStreamer<>(DataTypes.UNDEFINED);
```

Comment:
Couldn't this use the valueType of the argument instead of undefined?

Reply:
This is a good idea.